sympy/stats/stochastic_process_types.py

```diff
@@ -998,10 +1068,150 @@ def probability(self, condition, given_condition=None, evaluate=True, **kwargs):
+
+    expr: Sympy Expression
+        Expression containing Random Indexed Symbols with variable time stamps
+    condition: Relational/Boolean Expression
+        Expression containing time bounds of variable time stamps in expr
+
+    Examples
+    ========
+
+    >>> from sympy.stats.stochastic_process_types import get_timerv_swaps, PoissonProcess
+    >>> from sympy import symbols, Contains, Interval
+    >>> x, t, d = symbols('x t d', positive=True)
+    >>> X = PoissonProcess("X", 3)
+    >>> get_timerv_swaps(x*X(t), Contains(X(t), Interval.Lopen(0, 1)))
+    ([Interval.Lopen(0, 1)], {X(t): X(1)})
+    >>> get_timerv_swaps((X(t)**2 + X(d)**2), Contains(X(t), Interval.Lopen(0, 1))
+    ... & Contains(X(d), Interval.Ropen(1, 4))) # doctest: +SKIP
+    ([Interval.Ropen(1, 4), Interval.Lopen(0, 1)], {X(d): X(3), X(t): X(1)})
+
+    Returns
+    =======
+
+    intervals: list
+        List of Intervals/FiniteSet on which each time stamp is defined
+    rv_swap: dict
+        Dictionary mapping variable time Random Indexed Symbol to constant time
+        Random Indexed Variable
+
+    """
+
+    if not isinstance(condition, (Relational, Boolean)):
+        raise ValueError("%s is not a relational or combination of relationals"
+            % (condition))
+    expr_syms = list(expr.atoms(RandomIndexedSymbol))
+    if isinstance(condition, (And, Or)):
+        given_cond_args = condition.args
+    else: # single condition
+        given_cond_args = (condition, )
+    rv_swap = {}
+    intervals = []
+    for expr_sym in expr_syms:
+        for arg in given_cond_args:
+            if arg.has(expr_sym.key) and isinstance(expr_sym.key, Symbol):
+                intv = _set_converter(arg.args[1])
+                diff_key = intv._sup - intv._inf
+                if diff_key == oo:
+                    raise ValueError("%s should have finite bounds" % str(expr_sym.name))
+                elif diff_key == S.Zero: # has singleton set
+                    diff_key = intv._sup
+                rv_swap[expr_sym] = expr_sym.subs({expr_sym.key: diff_key})
+                intervals.append(intv)
+    return intervals, rv_swap
+
+
+class RandomWalk(DiscreteTimeStochasticProcess):
+
+    index_set = S.Naturals0
+
+    def __new__(cls, sym, p=S.Half, success=1, failure=-1,
+                            state_space=S.Integers):
+        sym = _symbol_converter(sym)
+        state_space = _set_converter(state_space)
+        p = _sympify(p)
+        success = _sympify(success)
+        failure = _sympify(failure)
+        _value_check(p > 0 and p <=1, 'p should be between 0 and 1.')
+        _value_check(success.is_integer, 'success should be an integer')
+        _value_check(failure.is_integer, 'fail should be an integer')
+        y = Function('f')
+        n = Symbol('n', integer=True)
+        f = y(n) - p*y(n + success) - (1 - p)*y(n + failure)
+        func = rsolve(f, y(n), {y(state_space._inf): 0, y(state_space._sup): 1})
+        func = Lambda(n, func)
```

Comment:
I think that the class args should match the given user args `obj.func(*obj.args)` so I don't think that the recurrence function should be wrapped in class creation, but rather be solved after creation for places where needed.

I also think that this is making the class totally unusable because it fails to solve when the state_space is set to infinite set.